### ISSUE TITLE
detect slow-to-compile code (and act accordingly)

### ISSUE THREAD

**Issue by fommil** [ensime/ensime-server#954]
My ENSIME server has been freezing up a lot recently, and I've also been using shapeless a lot more in my work projects.

This is a stack trace of what the PC is up to when it is completely non-responsive. Killing it and starting again is only a temporary solution because the same code will freeze everything.

@milessabin is there any way to halt early or something in the PC when using shapeless?

```
"Scala Presentation Compiler []" #108 daemon prio=5 os_prio=0 tid=0x00007fcd3406d000 nid=0x7004 runnable [0x00007fcd3e1c7000]
   java.lang.Thread.State: RUNNABLE
    at scala.runtime.ScalaRunTime$.hash(ScalaRunTime.scala:206)
    at scala.reflect.internal.util.HashSet.addOldEntry(HashSet.scala:81)
    at scala.reflect.internal.util.HashSet.growTable(HashSet.scala:101)
    at scala.reflect.internal.util.HashSet.addEntry(HashSet.scala:63)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$LocalShadower$1.addInfos(Implicits.scala:792)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$$anonfun$19.apply(Implicits.scala:873)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$$anonfun$19.apply(Implicits.scala:871)
    at scala.collection.immutable.List.flatMap(List.scala:327)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation.<init>(Implicits.scala:871)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:986)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.bestImplicit(Implicits.scala:1349)
    at scala.tools.nsc.typechecker.Implicits$class.inferImplicit(Implicits.scala:73)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.typedImplicit0(Implicits.scala:582)
<thousands of lines>

   Locked ownable synchronizers:
    - None
```

**Comment by fommil** [comment]
and what looks like the blocked ENSIME thread

```
"default-akka.actor.default-dispatcher-2" #21 prio=5 os_prio=0 tid=0x00007fce348ec000 nid=0x4f36 in Object.wait() [0x00007fcd833df000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at scala.tools.nsc.util.InterruptReq.getResult(InterruptReq.scala:39)
	- locked <0x0000000640d30cc0> (a scala.tools.nsc.util.WorkScheduler$$anon$1)
	at scala.tools.nsc.util.WorkScheduler.doQuickly(WorkScheduler.scala:58)
	at scala.tools.nsc.interactive.CompilerControl$class.ask(CompilerControl.scala:245)
	at scala.tools.nsc.interactive.Global.ask(Global.scala:104)
	at org.ensime.core.RichCompilerControl$class.askOption(RichPresentationCompiler.scala:29)
	at org.ensime.core.RichPresentationCompiler.askOption(RichPresentationCompiler.scala:187)
	at org.ensime.core.RichCompilerControl$class.askSymbolDesignationsInRegion(RichPresentationCompiler.scala:158)
	at org.ensime.core.RichPresentationCompiler.askSymbolDesignationsInRegion(RichPresentationCompiler.scala:187)
	at org.ensime.core.Analyzer.process(Analyzer.scala:242)
	at org.ensime.core.Analyzer$$anonfun$receive$1.applyOrElse(Analyzer.scala:90)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
	at org.ensime.core.Analyzer.aroundReceive(Analyzer.scala:26)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
	at akka.dispatch.Mailbox.run(Mailbox.scala:221)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by milessabin** [comment]
Not that I know of.

From the looking of the stacktrace it's probably happening during type class derivation ... have you looked into using the new `cachedImplicit` infrastructure to make sure derivations for a given type only happen once?

**Comment by fommil** [comment]
that sounds interesting, where can I read more about it?

**Comment by milessabin** [comment]
You will soon be able to subscribe to my newsletter. In the meantime see [here](https://github.com/milessabin/shapeless/blob/master/core/src/test/scala/shapeless/implicits.scala).

**Comment by fommil** [comment]
thanks! Feel free to unsubscribe from this ticket. We're going to action a minimal failing case and a workaround... because when this happens it basically wipes out ENSIME and there may be other dodgy user codes that can do the same.

**Comment by milessabin** [comment]
Okey doke ... let me know how you get on with `cachedImplicit` ... I'd be interested to hear if it has a beneficial effect on compile times.

**Comment by fommil** [comment]
@milessabin re `cachedImplicit` that actually sounds awesome. I've always been annoyed at the recomputation and I even have a suggestion in the spray-json stuff to define "explicit implicits" but maybe this will just fix it entirely.

**Comment by milessabin** [comment]
Hopefully you should be able to get away with just one cached implicit at the root of your hierarchy.

**Comment by fommil** [comment]
btw, I think it was this code that is causing the trouble, so possibly the basis of a test that clogs up the PC http://stackoverflow.com/a/29679342/1041691

```scala
sealed trait Traity
case object Foo extends Traity
case class Bar() extends Traity
case class Baz() extends Traity

import shapeless._

class NameHelper[A] {
  def apply[C <: Coproduct, K <: HList]()(implicit
    gen: LabelledGeneric.Aux[A, C],
    keys: ops.union.Keys.Aux[C, K],
    toSet: ops.hlist.ToTraversable.Aux[K, Set, Symbol]
  ): Set[String] = toSet(keys()).map(_.name)
}

def names[A] = new NameHelper[A]

val traities = names[Traity]()
```

**Comment by fommil** [comment]
I'm reopening this. It would be good to have support for blacklisting a file that is known to take a long time to compile. e.g. with a comment at the top. client or server side, that's the question...

**Comment by rorygraves** [comment]
Server

**Comment by fommil** [comment]
@sschaef what do you guys do when there is a file that takes a long time to compile? We'd like to exclude it from the load list with a comment in the source. It would be good if we both used the same comment, for example. And do you know any way of getting stats out of the presentation compiler? It would be good to detect which files were causing trouble and prompt the user to exclude them.

**Comment by fommil** [comment]
also, being able to cancel long-running presentation compiler loads would be great. But that would involve performance/timeout checks at appropriate parts in the presentation compiler code.

**Comment by sschaef** [comment]
We block the user until the compiler is finished, this is what we do. Right now we don't provide any functionality to disable the PC completely for entire files but that is on our todo list. For now, if the compiler takes too long, you are screwed.

I would not like a solution where users have to put a comment in the source (source pollution) in order to disable the PC, this should be configured in an external file, managed by the IDE.

Something else we do is to abort further waiting after a (configurable) timeout. But this is problematic because we can't interrupt the PC, we can only spawn it in the background and not wait for the result when the time is over. This leads in the worst case to a zombie PC that may be blocked for forever. We collect all the data by ourselves, I never even thought of getting such data from the compiler. I thought about a PC-pool where each PC runs in its own JVM. This way it would be possible to kill PCs, but that is not the solution that should be chosen. That the PC is non interruptible is a design problem of the compiler and there is not much we can do beside from fixing the compiler.

This reminds me that I wanted to ask Martin if dotdys PC is better engineered in that regard.

**Comment by aemoncannon** [comment]
Hmmm. Perhaps separate jvm per Analyzer would be manageable for ensime,
given the actor-based design?

**Comment by fommil** [comment]
We could now move to an `Analyzer` per module relatively easily, which has a lot of benefits, but actually killing a timed out PC would be great.

Hmm, spawing a new JVM... I'll have to think about that. A little bit of refactoring and that might just be possible. I guess it doesn't help much if all we do is spawn PCs in separate JVMs and then all they do is spin resolving implicits... we still need to know how to detect and blacklist the bad files.

Maybe a `.scala-pc-blacklist` file or something could work. But figuring out which file is causing the problem isn't always obvious.

**Comment by rorygraves** [comment]
Seperate JVMs was part of my thinking when I was talking about containerised services -  we even had a slide for it http://ensime.github.io/talks/scaladays15/#/8/6

We are moving to a separated enough world to actually do it.  But I think its a 2.0+ idea

**Comment by fommil** [comment]
maybe we should fork the presentation compiler and put some timeout / bail logic in there.

**Comment by rorygraves** [comment]
Thats not really a road I would like to go down right now.
We have enough crazy things to do, who knows what state the PC will be in if you hack interrupts?

**Comment by fommil** [comment]
:trollface:

**Comment by fommil** [comment]
Stats https://github.com/scala/scala/blob/1fbce4612c21a4d0c553ea489b4765494828c09f/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L5609

**Comment by fommil** [comment]
https://github.com/scala/scala/blob/1fbce4612c21a4d0c553ea489b4765494828c09f/src/reflect/scala/reflect/internal/Types.scala#L4658

**Comment by fommil** [comment]
https://github.com/scala/scala/blob/1fbce4612c21a4d0c553ea489b4765494828c09f/src/compiler/scala/tools/nsc/Global.scala#L229

**Comment by fommil** [comment]
https://github.com/scala/scala/blob/1fbce4612c21a4d0c553ea489b4765494828c09f/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L5417

**Comment by fommil** [comment]
@sschaef have you seen these stats hooks before? Just pasting lots of links that @xeno-by has told us about

**Comment by sschaef** [comment]
No I didn't know about them.

**Comment by fommil** [comment]
looking at the last one, looks like we might be able to use it to detect slow files! Then we can exclude them if the user agrees... perhaps we could agree on a common file format (or code comment) to ignore?

**Comment by retronym** [comment]
Suggestion: `global.currentRun.cancel()`

**Comment by fommil** [comment]
Jason also suggests instrumenting assertCorrectThread and Global#signalDone the former called a lot and the latter "The presentation compiler overrides that method to check if a targetted typecheck has been fulfilled and throws an exception to abort typechecking"

**Comment by weeshy** [comment]
I've had the same problem with developer ensime version. Few times per hour ensime hangs out with messages like that:

```
21:45:02.171 ERROR akka://ENSIME/user/ensime-main/project/indexer a.a.OneForOneStrategy - Futures timed out after [30 seconds]
java.util.concurrent.TimeoutException: Futures timed out after [30 seconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
	at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:190)
	at akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread$$anon$3.block(ThreadPoolBuilder.scala:169)
	at scala.concurrent.forkjoin.ForkJoinPool.managedBlock(ForkJoinPool.java:3640)
	at akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread.blockOn(ThreadPoolBuilder.scala:167)
	at scala.concurrent.Await$.result(package.scala:190)
	at org.ensime.indexer.SearchService.searchClasses(SearchService.scala:243)
	at org.ensime.core.Indexer.oldSearchTypes(Indexer.scala:28)
	at org.ensime.core.Indexer$$anonfun$receive$1.applyOrElse(Indexer.scala:53)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.event.LoggingReceive.apply(LoggingReceive.scala:62)
	at akka.event.LoggingReceive.apply(LoggingReceive.scala:50)
	at scala.PartialFunction$class.applyOrElse(PartialFunction.scala:123)
	at akka.event.LoggingReceive.applyOrElse(LoggingReceive.scala:50)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:467)
	at org.ensime.core.Indexer.aroundReceive(Indexer.scala:16)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
	at akka.dispatch.Mailbox.run(Mailbox.scala:220)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
After that ensime refuses to autocomplete code. The only workaround I had discovered is to restart the ensime. I could not factor out what code exactly causes this behaviour. I use few implicits and macroses in the project but without excess.

**Comment by vermest** [comment]
Hi,

Just my 2 cents...
Can this be the problem of the search service, where the future timeout is hard-coded:
`private val QUERY_TIMEOUT = 30 seconds`

And the search functions are not handling exceptions?:
`/** free-form search for classes */
  def searchClasses(query: String, max: Int): List[FqnSymbol] = {
    val fqns = Await.result(index.searchClasses(query, max), QUERY_TIMEOUT)
    Await.result(db.find(fqns), QUERY_TIMEOUT) take max
  }

  /** free-form search for classes and methods */
  def searchClassesMethods(terms: List[String], max: Int): List[FqnSymbol] = {
    val fqns = Await.result(index.searchClassesMethods(terms, max), QUERY_TIMEOUT)
    Await.result(db.find(fqns), QUERY_TIMEOUT) take max
  }
`
Seems that the caller Indexer service is not dealing with the exceptions either.

**Comment by fommil** [comment]
@vermest these are different issues, if you think there is some extra error handling you'd like to handle can you please create a PR for discussion (or at least a separate issue)?

**Comment by vermest** [comment]
OK, sure, I guess, I will look a bit deeper into the whole code before doing that. 
Although lately I'm pretty often affected by the similar timeout issues.
Otherwise editor works great, it's just that I have to reload the server ~3 times an hour.
Still faster than Eclipse or Idea ;)

**Comment by fommil** [comment]
symptom not cause

**Comment by fommil** [comment]
this could maybe be unblocked by https://github.com/scalacenter/scalac-profiling // @jvican 

It would be great to be able to report slow derivations as warnings / info through the reporter, and potentially opt to null out that code (e.g. the cachedImplicit hack)
